drivers: i2c: nrfx_twim: add support for device runtime PM

gmarull opened the pull request:
Enable device runtime PM on the I2C driver. Note that this mechanism
replaces the old nrfx_twim_enable/disable calls with
pm_device_runtime_get/put, which means that driver will not save power
unless CONFIG_PM_DEVICE_RUNTIME=y.

Some quick measurements on thingy_52 running the samples/sensor/hts221
show a decrease of ~2-3uA in average when enabling device runtime PM.
Note that the driver already had implicit PM before, so the change for
users will be ~none, except that they now must enable the PM subsystem
features. While this case is not the best example, the PM subsystem will
turn to be beneficial as a whole when all devices in the board implement
it.

I've also done some minor cleanups as prep-work commits, please check the details

gmarull committed:
drivers: i2c: nrfx_twim: drop twim_initialized usage

The driver re-initialized itself on the twim_transfer call if it wasn't.
This condition was likely added because of some manual/custom PM
schemes. Drop it in preparation for runtime PM.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: i2c: nrfx_twim: remove redundant deinit_twim shim

The shim had no value, just call nrfx_twim_uninit directly.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: i2c: nrfx_twim: enable/disable when resuming/suspending

The driver already implemented some sort of runtime PM by
unconditionally calling nrfx_twim_disable/enable. Perform the same
operations in the PM callback instead of doing a full init/deinit every
time.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: i2c: nrfx_twim: enable device runtime PM

Enable device runtime PM on the I2C driver. Note that this mechanism
replaces the old nrfx_twim_enable/disable calls with
pm_device_runtime_get/put, which means that driver will not save power
unless CONFIG_PM_DEVICE_RUNTIME=y.

Some quick measurements on thingy_52 running the samples/sensor/hts221
show a decrease of ~2-3uA in average when enabling device runtime PM.
Note that the driver already had implicit PM before, so the change for
users will be ~none, except that they now must enable the PM subsystem
features. While this case is not the best example, the PM subsystem will
turn to be beneficial as a whole when all devices in the board implement
it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: i2c: nrfx_twim: create common init function

The device definition macro defined per-device init function, resulting
in a cluttered and hard to extend init function. Create a per-instance
IRQ connect function instead, which is called at init time by a common
function.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: i2c: nrfx_twim: drop init_twim shim

The function was just a shim to nrfx_twim_init, call it directly
instead.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

carlocaione commented on drivers/i2c/i2c_nrfx_twim.c:
```diff
@@ -344,6 +330,8 @@ static int twim_nrfx_pm_action(const struct device *dev,
 {
 	const struct i2c_nrfx_twim_config *dev_config = dev->config;
+	struct i2c_nrfx_twim_data *dev_data = dev->data;
+	nrfx_err_t nerr;
```
unused?

anangl commented on drivers/i2c/i2c_nrfx_twim.c:
```diff
@@ -52,19 +51,12 @@ static int i2c_nrfx_twim_transfer(const struct device *dev,
 	k_sem_take(&dev_data->completion_sync, K_NO_WAIT);
 
-	nrfx_twim_enable(&dev_config->twim);
+	(void)pm_device_runtime_get(dev);
```
When `nrfx_twim_bus_recover` is called, the peripheral needs to be disabled (so that the pins are released) and in the current form this is not the case.
Actually, couldn't the calls to `nrfx_twim_enable/nrfx_twim_disable` stay here and instead be removed from `twim_nrfx_pm_action` (is that a problem if the shim reduces power consumption even if not instructed to)?

gmarull committed:
drivers: i2c: nrfx_twim: avoid init/deinit to update frequency

nrfx HAL does not support updating frequency at runtime, which means we
need to do a full init/deinit to update it when it is a matter of a
simple register write. Fix this by using the lower level HAL.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: i2c: nrfx_twim: disable and restore state for bus recovery

TWIM peripheral needs to be disabled for bus recovery, otherwise SCL/SDA
lines will not be released. This patch makes sure to disable peripheral
if active, and, restore its state afterwards (including pin
configuration).

It is worth to note that a better solution would be to:

1. Define scl/sda pins as `-gpios` in DT
2. Use GPIO API in the driver to perform recovery (as some other drivers
   do)
3. Potentially use a "gpio" pinctrl state for this case

Unfortunately HAL is doing everything under the hood, so we have little
options to improve this unless we don't use it for such case. GPIO based
recovery should likely be generalized as many drivers seem to replicate
such _algorithm_.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on drivers/i2c/i2c_nrfx_twim.c:
```diff
@@ -22,7 +23,6 @@ struct i2c_nrfx_twim_data {
 	struct k_sem transfer_sync;
 	struct k_sem completion_sync;
 	nrfx_twim_config_t twim_config;
-	bool twim_initialized;
```
I have switched to `nrf_twim_frequency_set`. For the `E0E1` case, we could introduce a pinctrl custom state instead, this is actually one of the usecases for states.

gmarull committed:
pm: device: provide dummy pm_device_state_get if CONFIG_PM_DEVICE=n

When CONFIG_PM_DEVICE=n a device will always be in active state. Provide
a dummy implementation of pm_device_state_get that always returns
PM_DEVICE_STATE_ACTIVE.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

anangl commented on drivers/i2c/i2c_nrfx_twim.c:
```diff
@@ -239,79 +231,47 @@ static void event_handler(nrfx_twim_evt_t const *p_event, void *p_context)
 	}
 
-	if (frequency != dev_data->twim_config.frequency) {
-		dev_data->twim_config.frequency = frequency;
```
Since `twim_config` is not changed now, it should be moved to the `i2c_nrfx_twim_config` structure.

gmarull replied:
good point

gmarull committed:
drivers: i2c: nrfx_twim: constify twim_config

The twim_config structure is no longer modified at runtime, so it can be
placed in driver's config (const).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
